### PR TITLE
[MI-2324]:Fixed review fixes of confluence auth/notification PR

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
Link to PR:- https://github.com/mattermost/mattermost-plugin-confluence/pull/89

### CODE REVIEW

File: server/user.go [hunk 1]
```diff
@@ -241,9 +241,9 @@ func (p *Plugin) CompleteOAuth2(mattermostUserID, code, state string, instance I
 
 	// Fetch the cloudid for cloud instance if not already present
 	if instance.Common().Type == CloudInstanceType && instance.(*cloudInstance).CloudID == "" {
-		cloudID, err := client.(*confluenceCloudClient).GetCloudID()
-		if err != nil {
-			return nil, nil, err
+		cloudID, CErr := client.(*confluenceCloudClient).GetCloudID()
```
Comment (ayusht2810) [inline server/user.go]:
```suggestion
		cloudID, cErr := client.(*confluenceCloudClient).GetCloudID()
```

File: server/service/edit_subscription.go [hunk 2]
```diff
@@ -18,9 +18,9 @@ func EditSubscription(subscription serializer.Subscription) error {
 	}
 	key := store.GetSubscriptionKey()
 	err = store.AtomicModify(key, func(initialBytes []byte) ([]byte, error) {
-		subscriptions, err := serializer.SubscriptionsFromJSON(initialBytes)
-		if err != nil {
-			return nil, err
+		subscriptions, SErr := serializer.SubscriptionsFromJSON(initialBytes)
```
Comment (ayusht2810) [inline server/service/edit_subscription.go]: same

File: server/save_subscription.go [hunk 3]
```diff
@@ -65,9 +65,9 @@ func (p *Plugin) handleSaveSubscription(w http.ResponseWriter, r *http.Request)
 		}
 
 		spaceKey := subscription.(*serializer.SpaceSubscription).GetSubscription().SpaceKey
-		resp, err := client.GetSpaceData(spaceKey)
-		if err != nil {
-			p.LogAndRespondError(w, http.StatusBadRequest, "Error getting space related data for space subscription.", err)
+		resp, GErr := client.GetSpaceData(spaceKey)
```
Comment (ayusht2810) [inline server/save_subscription.go]: same

File: server/kv.go [hunk 4]
```diff
@@ -410,9 +410,9 @@ func (store store) DeleteConnection(instanceID, mattermostUserID types.ID) (retu
 
 	// Check for whether the admin token stored for each confluenceURL is of the current user or not. If it is then delete that admin connection also
 	if c.IsAdmin {
-		adminConnection, err := store.LoadConnection(instanceID, AdminMattermostUserID)
-		if err != nil {
-			return err
+		adminConnection, LErr := store.LoadConnection(instanceID, AdminMattermostUserID)
```
Comment (ayusht2810) [inline server/kv.go]: same

File: server/client_server.go [hunk 5]
```diff
@@ -141,36 +141,27 @@ func (csc *confluenceServerClient) CreateWebhook(subscription serializer.Subscri
 		},
 	}
 	webhookResponse := &WebhookResponse{}
-	url, err := utils.GetEndpointURL(csc.URL, PathCreateWebhook)
-	if err != nil {
-		return nil, errors.Wrap(err, "confluence CreateWebhook")
-	}
-	_, err = utils.CallJSON(csc.URL, http.MethodPost, url, requestBody, webhookResponse, csc.HTTPClient)
+
```
Comment (ayusht2810) [inline server/client_server.go]:
```suggestion
```

File: server/client_cloud.go [hunk 6]
```diff
@@ -167,13 +150,10 @@ func (ccc *confluenceCloudClient) GetSpaceData(spaceKey string) (*SpaceResponse,
 
 func (ccc *confluenceCloudClient) GetUserGroups(connection *Connection) ([]*UserGroup, error) {
 	userGroups := UserGroups{}
-	url, err := utils.GetEndpointURL(ccc.URL, fmt.Sprintf(PathGetUserGroupsForCloud, connection.AccountID))
-	if err != nil {
-		return nil, errors.Wrap(err, "confluence GetUserGroups")
-	}
-	_, err = utils.CallJSON(ccc.URL, http.MethodGet, url, nil, &userGroups, ccc.HTTPClient)
+
```
Comment (ayusht2810) [inline server/client_cloud.go]:
```suggestion
```

File: server/client_cloud.go [hunk 7]
```diff
@@ -123,41 +115,32 @@ func (ccc *confluenceCloudClient) GetEventData(confluenceCloudEvent *serializer.
...
 }
 
 func (ccc *confluenceCloudClient) GetSpaceData(spaceKey string) (*SpaceResponse, error) {
 	spaceResponse := &SpaceResponse{}
-	url, err := utils.GetEndpointURL(ccc.URL, fmt.Sprintf(PathSpaceData, spaceKey))
-	if err != nil {
-		return nil, errors.Wrap(err, "confluence GetSpaceData")
-	}
-	_, err = utils.CallJSON(ccc.URL, http.MethodGet, url, nil, spaceResponse, ccc.HTTPClient)
+
```
Comment (ayusht2810) [inline server/client_cloud.go]:
```suggestion
```

File: server/edit_subscription.go [hunk 8]
```diff
@@ -46,46 +47,72 @@ func (p *Plugin) handleEditChannelSubscription(w http.ResponseWriter, r *http.Re
...
+	}
+
+	if err := p.DeleteSubscription(subscription.GetChannelID(), oldSubscription.GetAlias(), userID); err != nil {
+		p.LogAndRespondError(w, http.StatusBadRequest, "not able to delete subscription.", err)
+		return
+	}
+
+	if instance.Common().Type == ServerInstanceType {
+		if err := p.CreateWebhook(instance, subscription, userID); err != nil {
+			p.LogAndRespondError(w, http.StatusBadRequest, "Not able to create webhook.", err)
```
Comment (ayusht2810) [inline server/edit_subscription.go:63]:
```suggestion
			p.LogAndRespondError(w, http.StatusBadRequest, "not able to create webhook.", err)
```

Comment (ayusht2810) [reply server/edit_subscription.go:63]: it is present in small letters in the above line

File: server/edit_subscription.go [hunk 9]
```diff
@@ -46,46 +47,72 @@ func (p *Plugin) handleEditChannelSubscription(w http.ResponseWriter, r *http.Re
...
-			return
+			return nil, nil, "Error decoding request body for space subscription.", http.StatusBadRequest, err
 		}
 
 		spaceKey := subscription.(*serializer.SpaceSubscription).GetSubscription().SpaceKey
-		resp, err := client.GetSpaceData(spaceKey)
-		if err != nil {
-			p.LogAndRespondError(w, http.StatusBadRequest, "Error getting space related data for space subscription.", err)
-			return
+		resp, GErr := client.GetSpaceData(spaceKey)
```
Comment (ayusht2810) [inline server/edit_subscription.go]:
```suggestion
		resp, gErr := client.GetSpaceData(spaceKey)
```

File: server/edit_subscription.go [hunk 10]
```diff
@@ -46,46 +47,72 @@ func (p *Plugin) handleEditChannelSubscription(w http.ResponseWriter, r *http.Re
...
-			p.LogAndRespondError(w, http.StatusBadRequest, "Error decoding request body for page subscription.", err)
-			return
+			return nil, nil, "Error decoding request body for page subscription.", http.StatusBadRequest, err
 		}
 
-		pageID, err := strconv.Atoi(subscription.(*serializer.PageSubscription).GetSubscription().PageID)
-		if err != nil {
-			p.LogAndRespondError(w, http.StatusInternalServerError, "Error converting pageID to integer.", err)
-			return
+		pageID, SErr := strconv.Atoi(subscription.(*serializer.PageSubscription).GetSubscription().PageID)
```
Comment (ayusht2810) [inline server/edit_subscription.go]: same

File: server/confluence_server.go [hunk 11]
```diff
@@ -31,7 +31,8 @@ func (p *Plugin) handleConfluenceServerWebhook(w http.ResponseWriter, r *http.Re
 		return
 	}
 
-	if status, err := verifyWebHookSecret(p.conf.Secret, r.Header.Get("X-Hub-Signature"), body); err != nil {
+	status, err := verifyWebHookSecret(p.conf.Secret, r.Header.Get("X-Hub-Signature"), body)
```
Comment (raghavaggarwal2308) [inline server/confluence_server.go]: Constant for `"X-Hub-Signature"`

File: server/edit_subscription.go [hunk 12]
```diff
@@ -46,46 +47,72 @@ func (p *Plugin) handleEditChannelSubscription(w http.ResponseWriter, r *http.Re
 		return
 	}
 
-	body, err := ioutil.ReadAll(r.Body)
+	subscription, oldSubscription, statusLog, statusCode, err := p.SubscriptionsFromJSON(r.Body, client, subscriptionType, oldSubscriptionType, userID)
 	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
+		p.LogAndRespondError(w, statusCode, statusLog, err)
```
Comment (raghavaggarwal2308) [inline server/edit_subscription.go]: Don't we need to return from here?

**Review by raghavaggarwal2308 — APPROVED** [review]

**Review by ayusht2810 — APPROVED** [review]

**Review by raghavaggarwal2308 — APPROVED** [review]

File: go.mod [hunk 13]
```diff
@@ -13,4 +13,5 @@ require (
 	github.com/thoas/go-funk v0.6.0
 	go.uber.org/atomic v1.9.0
 	golang.org/x/oauth2 v0.0.0-20211104180415-d3ed0bb246c8
+	golang.org/x/sys v0.0.0-20220928140112-f11e5e49a4ec // indirect
```
Comment (manojmalik20) [inline go.mod]: I think you can remove indirect dependencies from here.

File: server/client_cloud.go [hunk 14]
```diff
@@ -167,13 +149,9 @@ func (ccc *confluenceCloudClient) GetSpaceData(spaceKey string) (*SpaceResponse,
...
 	userGroups := UserGroups{}
-	url, err := utils.GetEndpointURL(ccc.URL, fmt.Sprintf(PathGetUserGroupsForCloud, connection.AccountID))
+	_, err := utils.CallJSONWithURL(ccc.URL, fmt.Sprintf(PathGetUserGroupsForCloud, connection.AccountID), http.MethodGet, nil, &userGroups, ccc.HTTPClient)
 	if err != nil {
-		return nil, errors.Wrap(err, "confluence GetUserGroups")
-	}
-	_, err = utils.CallJSON(ccc.URL, http.MethodGet, url, nil, &userGroups, ccc.HTTPClient)
-	if err != nil {
-		return nil, errors.Wrap(err, "confluence GetUserGroups")
+		return nil, errors.Wrap(err, "confluence GetSpaceData")
```
Comment (manojmalik20) [inline server/client_cloud.go]: I think the error is wrong here.

File: server/edit_subscription.go [hunk 15]
```diff
@@ -46,46 +47,73 @@ func (p *Plugin) handleEditChannelSubscription(w http.ResponseWriter, r *http.Re
 		return
 	}
 
-	body, err := ioutil.ReadAll(r.Body)
+	subscription, oldSubscription, statusLog, statusCode, err := p.SubscriptionsFromJSON(r.Body, client, subscriptionType, oldSubscriptionType, userID)
```
Comment (manojmalik20) [inline server/edit_subscription.go]:
```suggestion
	subscription, oldSubscription, logMessage, statusCode, err := p.SubscriptionsFromJSON(r.Body, client, subscriptionType, oldSubscriptionType, userID)
```

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**